logstash-core/lib/logstash/config/mixin.rb

```diff
@@ -126,6 +146,42 @@ def config_init(params)
     @config = params
   end # def config_init
 
+  # Replace all environment variable references in 'value' param by environment variable value and return updated value
+  # Process following patterns : $VAR, ${VAR}, ${VAR:defaultValue}
+  def replace_env_placeholders(value)
+    if (value.is_a?(String))
+      while value =~ ENV_PLACEHOLDER_REGEX
+        valueParts = value.partition(ENV_PLACEHOLDER_REGEX)
+        placeHolder = valueParts[1]
+        if placeHolder.start_with?("${")
+          envVarName = placeHolder.slice(2..placeHolder.length - 2)
+        else
+          envVarName = placeHolder.slice(1..placeHolder.length - 1)
+        end
+        if envVarName.include?(':')
+          placeHolderParts = envVarName.split(':')
+          envVarName = placeHolderParts[0]
+          defaultValue = placeHolderParts[1] || ""
+        else
+          defaultValue = ""
+        end
+        
+        envVarValue = ENV[envVarName]
+        if envVarValue.nil? 
+          envVarValue = defaultValue
+          if defaultValue.empty?
+            @logger.warn("In plugin '#{self.class.config_name}', " +
+                         "referenced environment variable '#{envVarName}' does not exist. " +
+                         "This reference has been replaced by empty string.")
+          end
+        end
+        value = valueParts[0] << envVarValue << valueParts[2]
+        @logger.info("Replacing config environment variable '#{placeHolder}' with #{envVarValue}")
```

Comment:
Yes, it would be lovely to show a real world example. The ones users had requested in #3944. Like injecting `TCP_PORT` or adding `tags` using environment variable